Bisect start

pranitbauva1997 opened the pull request:
@chriscool @larsxscheider I am stuck on this for quite some time now and it would be very helpful if you could look into this. Two problems are left to address
1. In tests which start with `git bisect reset`, it shows an error. I have tried to do a local git installation by `make -j 4 prefix=~/git install V= "CFFLAGS=-O0 -ggdb3" and then using the binary`~/git/bin/git` to reproduce the statements in the test but I don't know for some reason it works with that but not in the test suite.
2. The `--term-bad` `--term-good` argument handling problem

chriscool commented:
Right now it looks like it doesn't compile:

```
builtin/bisect--helper.c:557:39: error: passing 'char *' to parameter of type 'const unsigned char *' converts between pointers to integer types with different sign [-Werror,-Wpointer-sign]

                if (update_ref(NULL, "BISECT_HEAD", start_head.buf, NULL, 0,

                                                    ^~~~~~~~~~~~~~

./refs.h:463:30: note: passing argument to parameter 'new_sha1' here

               const unsigned char *new_sha1, const unsigned char *old_sha1,

                                    ^

1 error generated.
```

pranitbauva1997 commented:
@chriscool for some reason it wasn't detected by my compiler. I got to know it by travis-ci. I have updated the code and it fixes it.

Also I am able to fix the 1st error. It was because I was only writing out the head which starts with "refs/head/" but it can also contain just an hash. So it is now fixed by having an extra else.

chriscool commented:
Great!

chriscool commented on builtin/bisect--helper.c:
```diff
@@ -404,6 +404,228 @@ static int bisect_terms(struct bisect_terms *terms, int term_defined)
+	 */
+	head = resolve_ref_unsafe("HEAD", 0, sha1, &flag);
+	if (!head) {
+		if(get_sha1("HEAD", sha1)) {
```
A space is missing after `if`.

chriscool commented:
As Junio suggested using OPT_CMDMODE for these arguments for --bisect-terms, I think this should be discussed on the list with him. He will certainly find the best way.

larsxschneider commented:
@pranitbauva1997 looks like t6030 has still failures. are these related to `--term-good` and `--term-bad` that you mentioned above?

pranitbauva1997 commented:
@larsxschneider Yes they are related only to option handling. I have eliminated others.

larsxschneider commented:
OK, can you reproduce the test result on your local machine?

pranitbauva1997 commented:
@larsxschneider [Here](http://paste.ubuntu.com/21252069/) is the normal output. [Here](http://paste.ubuntu.com/21252140/) is the verbose output

chriscool commented:
I think the first error is this:

```
expecting success: 
	git bisect reset &&
	git bisect start HEAD $HASH1 -- dir1 &&
	para1=$(git rev-parse --verify HEAD) &&
	test "$para1" = "$PARA_HASH1" &&
	git bisect bad > my_bisect_log.txt &&
	grep "$PARA_HASH1 is the first bad commit" my_bisect_log.txt

We are not bisecting.
fatal: 'dir1' does not appear to be a valid revision
not ok 39 - restricting bisection on one dir
#	
#		git bisect reset &&
#		git bisect start HEAD $HASH1 -- dir1 &&
#		para1=$(git rev-parse --verify HEAD) &&
#		test "$para1" = "$PARA_HASH1" &&
#		git bisect bad > my_bisect_log.txt &&
#		grep "$PARA_HASH1 is the first bad commit" my_bisect_log.txt
#	
```

When `git bisect start HEAD $HASH1 -- dir1` it looks like your code does as if `dir1` was before `--`.
But anything after `--` should not be considered revision, it is a path.

chriscool commented:
There is a reason why the shell code first scans all the args to see if one of them is `--`:

```
-	has_double_dash=0
-	for arg; do
-		case "$arg" in --) has_double_dash=1; break ;; esac
-	done
```

it looks like the C code doesn't do the same thing.

pranitbauva1997 commented:
@chriscool Yes! Thanks for explaining.

@larsxschneider parse_options() "reads" the "--" and pushes it at the back of argv so the C code can't read the "--" and thus it treats dir1 as a normal commit.

chriscool commented on builtin/bisect--helper.c:
```diff
@@ -404,6 +404,228 @@ static int bisect_terms(struct bisect_terms *terms, int term_defined)
 	return 0;
```
It looks like --bisect-start is in the synopsis at the beginning of this file.

pranitbauva1997 replied:
Will add it. Thanks!

chriscool commented on builtin/bisect--helper.c:
```diff
@@ -404,6 +404,228 @@ static int bisect_terms(struct bisect_terms *terms, int term_defined)
+
+	if (is_bare_repository())
+		no_checkout = 1;
+
```
I think we should check if there is a double dash in a separate loop, before the loop below, as the shell does, otherwise the behavior will not be the same.

pranitbauva1997 replied:
Okay sure. I will check it as it is done in the shell code.

chriscool commented on builtin/bisect--helper.c:
```diff
@@ -404,6 +404,228 @@ static int bisect_terms(struct bisect_terms *terms, int term_defined)
+			string_list_clear(&states, 0);
+			die(_("unrecognised option: '%s'"), argv[i]);
+		}
+		if (get_oid(argv[i], &oid) || has_double_dash) {
```
Here I think we need to know if there is a double dash after the current arg.

pranitbauva1997 replied:
True!

pranitbauva1997 replied:
For that, I think omitting the initial check for has_double_dash is more appropriate but to include it in the argument handling for loop.

pranitbauva1997 commented:
@chriscool @larsxschneider I will send out my work on this after your and Junio's review tomorrow.

chriscool commented:
There is the following build error:

```
builtin/bisect--helper.c: In function ‘bisect_next_check’:
builtin/bisect--helper.c:369:2: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]
builtin/bisect--helper.c: In function ‘bisect_start’:
builtin/bisect--helper.c:599:3: error: implicit declaration of function ‘sq_quote_argv’ [-Werror=implicit-function-declaration]
cc1: all warnings being treated as errors
make: *** [builtin/bisect--helper.o] Error 1
make: *** Waiting for unfinished jobs....
```

larsxschneider commented on builtin/bisect--helper.c:
```diff
@@ -428,6 +431,217 @@ static int bisect_terms(struct bisect_terms *terms, int term_defined)
+	 * set references named with the default terms, and won't be able
+	 * to change afterwards.
+	 */
+	must_write_terms = !!revs.nr;
```
You might want to avoid this `!!` trick... see Documentation/CodingGuidelines

chriscool replied:
It says:

```
Some clever tricks, like using the !! operator with arithmetic constructs, can be extremely confusing to others.
```

I think the "with arithmetic constructs" part is important to understand this guideline, as the trick is used in 114 places already in the code.

So I think here it is ok.

pranitbauva1997 replied:
This was suggested by Junio.

larsxschneider commented on builtin/bisect--helper.c:
```diff
@@ -428,6 +431,217 @@ static int bisect_terms(struct bisect_terms *terms, int term_defined)
 	return 0;
 }
 
+static int bisect_start(struct bisect_terms *terms, int no_checkout,
```
minor: I am often confused by `no_...` variables as I think they are harder to read. Usually it is easer to flip the variable and add a `!`. Down below you use `!no_checkout` which would then be just `checkout`...

pranitbauva1997 replied:
The git-bisect.sh and bisect.c uses no_checkout. And its probably named no_checkout because by default we do checkout, but if in some circumstances we can use no_checkout

chriscool commented:
There is still a build error:

```
builtin/bisect--helper.c: In function ‘bisect_next_check’:
builtin/bisect--helper.c:370:2: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]
cc1: all warnings being treated as errors
make: *** [builtin/bisect--helper.o] Error 1
make: *** Waiting for unfinished jobs....
```

pranitbauva1997 commented:
@chriscool Its now solved.

pranitbauva1997 commented:
@chriscool @larsxschneider I have FINALLY finished porting bisect_start(). Could you please review it once? I will now start with bisect_next()
